enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,209 @@
+
+#### Operand cleanup status
+
+To provide visibility on the status of operand cleanup, OLM can list out the CRs that are still pending deletion in the CSV status e.g `status.cleanup.pendingDeletion`.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+status:
+  ...
+  cleanup:
+    pendingDeletion:
+    - resource: etcdclusters.etcd.database.coreos.com
+      kind: EtcdCluster
+      version: v1beta2
+      name: Foo
+      namespace: Bar
+    - kind: EtcdCluster
+      ...
+```
+
+This information can be useful when debugging a stalled operator uninstall process to see the CRs whose removal is blocking operand cleanup.
+
+**NOTE:** The number of CRs that OLM has to list in the status is unbounded and could potentially cause the CSV size to exceed the [max request size][max-request-size] configured in etcd. For this reason there would have to be an upper limit on the number of CRs displayed in `status.cleanup.pendingDeletion`.
+
+#### Handling CSV deletion due to replacement
+
+A CSV can also be [deleted by OLM][csv-deletion-on-upgrade] during it's lifecycle when OLM detects a newer CSV that replaces it as part of an upgrade. For an operator that has already opted-in to operand cleanup, this would result in OLM cleaning up all the CRs before removing the old CSV.
+
+To prevent the cleanup of CRs on an upgrade, OLM should ensure that a CSV which has `spec.cleanup.enabled: true` and is marked for replacement (`phase: replacing`), should first be cleared of the `olm/operand-cleanup` finalizer and opted-out of cleanup before it is marked for deletion or deleted.
+
+#### CRD removal and caveats
+
+While not strictly necessary for the goal of cleaning up operator-created resources, the cleanup process can also include removing the CRDs after the CRs and operator have been removed. There are a few edge cases consider before removing the CRD as well.
+
+**Namespaced scoped operators:**
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
+
+- The operator installed is global
+- The non-global operator is the only installation for that operator type in the cluster, i.e no other CSVs provide the same owned APIs.
+
+**Resolver cache:**
+
+OLM caches all the on-cluster subscriptions, catalogs, and CSVs to determine what the provided APIs are for dependency resolution (TODO: citation/link needed). It is possible that OLM could install an unusable operator if another operator that provides the required API is uninstalled from the cluster and the resolver cache is stale. The cache should be invalidated in the event of a CRD cleanup on operator uninstall.
```

Comment:
> OLM caches all the on-cluster subscriptions, catalogs, and CSVs to determine what the provided APIs are for dependency resolution

The cache is only for the contents of catalogs. It doesn't retain CSVs or Subscriptions.

Reply:
Marking as resolved. Removed this section.